domains/DecisionTree.ml

```diff
@@ -1317,7 +1300,97 @@ struct
         let b = match domain with | None -> B.inner env vars cs | Some domain -> B.meet (B.inner env vars cs) domain in
         if B.isBot b then () else Format.fprintf fmt "%a ? %a\n" B.print b F.print f
       | Node((c,nc),l,r) -> aux r (nc::cs); aux l (c::cs)
-    in aux t.tree []; Format.fprintf fmt "\nDOMAIN = {%a}%a\n" print_domain domain (print_tree vars) t.tree
+    (* in aux t.tree []; Format.fprintf fmt "\nDOMAIN = {%a}%a\n" print_domain domain (print_tree vars) t.tree *)
+    (* Format.fprintf fmt "\nDOMAIN = {%a}%a\n" print_domain domain (print_tree vars) t.tree; *)
+    in aux t.tree []
+
+
+  (* 
+     Takes left and right tree as argument and cuts away all parts of the left tree
+     that are not part of the domain of the righ tree.
+
+     This means that if some part of the domain of the right tree is undefined (i.e. bottom, top or NIL)
+     then the corresponding part in the left tree is replaced with a bottom leaf.
+
+     NOTE: narrow_left is only monotone w.r.t. the APPROXIMATION order
+  *)
+  let left_narrow t_left t_right =
+    let domain = t_left.domain in 
+    let env = t_left.env in 
+    let vars = t_left.vars in 
+    let botLeaf = Leaf (F.bot env vars) in
+    let isDefined f = not (F.isBot f || F.isTop f) in
+    let fBotLeft _ _ = Bot in (* LHS is bottom, keep it that way *)
+    let fBotRight _ fLeft = if isDefined fLeft then botLeaf else Leaf fLeft in (* if RHS is NIL and LHS is defined then go to bottom *)
+    let fLeaf cs l1 l2 = 
+      if isDefined l2 then Leaf l1 (* don't change if RHS is defined*)
+      else (* if RHS is not defined, then go to bottom if LHS is not already top or bottom*)
+        if isDefined l1 then botLeaf 
+        else Leaf l1
+    in { 
+      domain = domain; 
+      tree = tree_join_helper fBotLeft fBotRight fLeaf t_left.tree t_right.tree env vars; 
+      env = env; 
+      vars = vars 
+    }
+
+
+  (*
+     Combination of reset and filter operation. This operator can be used to implement the CTL 'until' operator.
+
+     Takes a decision trees 't_keep', 't_reset' and 't' an as argument and resets all leafs in 't' 
+     that are also part of the domain of 't_reset' and removes all leafs in 't' that are neither part of the domain of 't_keep' or 't_reset'.
+
+
+     The function first filters out all leafs in 't' that are not also part of the domain of 't_keep' and 't_reset'. 
+     Then it resets all leafs in 't' that are also part of the domain of 't_reset'.
+  
+  *)
+  let reset_until t_keep t_reset t =
```

Comment:
I changed it accordingly. 

The reason why I had it this way was to make partial application with t_keep and t_reset more straight forward. But I agree that consistency with the other functions is more important.